HPCC-30411 Add support for dynamically updating TLS config

ghalliday opened the pull request:
<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [ ] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

ghalliday committed:
HPCC-30411 Add support for dynamically updating TLS config
HPCC-30754 Allow roxie to use issuer based tls in bare-metal configuration

Signed-off-by: Gavin Halliday <gavin.halliday@lexisnexis.com>

afishbeck commented on system/jlib/jsecrets.cpp:
```diff
@@ -1349,18 +1402,10 @@ IPropertyTree *createIssuerTlsClientConfig(const char *issuer, bool acceptSelfSi
 }
 
-IPropertyTree *getIssuerTlsServerConfig(const char *name)
+static ISyncedPropertyTree * createIssuerTlsServerConfig(const char *name)
```
I know it's existing code (but since you are changing these).. I've been slowly changing the "name" parameter on these routines to "issuer" to be more self documenting.  Or if you prefer "issuerName".

The issuer which is part of the path to the secret act like categories for the various trust zones: public, local, remote, etc.

afishbeck commented on system/jlib/jsecrets.cpp:
```diff
@@ -1441,21 +1501,18 @@ jlib_decl bool queryMtls()
+                        if (!isEmptyString(passPhrase))
+                            info->setProp("passphrase", passPhrase); // encrypted
+
+                        Owned<ISyncedPropertyTree> entry = createSyncedPropertyTree(info);
```
Not really new, although changing, but there may be a timing issue here because the meaning of "local" is being overridden.

In practical terms, looking up "local" means, get tls configuration based on the secret at "/opt/HPCCsystems/secrets/certificates/local".

The various ways of configuring bare metal need consideration.
One way of configuring bare metal could be to put TLS info into the secret locations we use in kubernetes. 
The other way of configuring bare-metal would be to fill in the custom TLS configuration values.

Here, a tree based on the custom bare-metal configuration values are turned into a tree that over rides the contents at the secret location.
If we ever (accidently?) use both the cached tree would be whichever came first.
So I guess we need to make sure if we are going to override "local" we always override "local".

afishbeck replied:
The alternative would be to cache under a different name, a name implying "bare-metal-mtls".  But we'd have to track down what was currently depending on it being called "local".

afishbeck commented on system/jlib/jsecrets.cpp:
```diff
@@ -1308,6 +1250,117 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+        config.setown(createPTree(issuer));
+
+        //GH-Tony should this condition be removed?
+        if (strieq(issuer, "remote")||strieq(issuer, "local"))
```
It looks like you've mixed some "ClientConfig" lines in with "ServerConfig" lines.
When inside ClientConfig these lines imply that we don't currently setup issuer client certificates for issuers other than remote and local.
Using this logic on the ServerConfig won't work, because server side we always want the certs included.

afishbeck commented on system/jlib/jptree.hpp:
```diff
@@ -429,4 +429,25 @@ extern jlib_decl StringBuffer &getExpertOptString(const char *opt, StringBuffer
+extern jlib_decl unsigned getPropertyTreeHash(const IPropertyTree & source, unsigned hashcode);
+
+//Interface for encapsulating an IPropertyTree that can be atomically updated.  The result of getTree() is guaranteed
+//to not be modified and to remain vali and consistent until it is released.
```
"valid"

afishbeck commented:
@ghalliday I've added some comments, but the general direction looks good.

Next steps, I think we want to:

1. Have getIssuerTlsConfig() and getIssuerTlsServerConfigWithTrustedPeers() return ISyncedPropertyTrees instead of regular IPropertyTrees.
2. We want to change CSecureSocketContext(const IPropertyTree* config, SecureSocketType sockettype) to take an ISyncedProtpertyTree instead of IPropertyTree... and to update the certs, keys, and CACerts if anything changes before a secure accept is performed.
3. Then clean up any pathway between those two functions.. or any calls to the CSecureSocketContext(IPropertyTree *) constructor that don't come from secrets (for example a PTree is built from bare-metal config).

When those code paths are cleaned up we should start to see any complications that might be there.

Be careful with the idea of caching the trusted_peers because it is not tied directly to the issuer so needs another level of cache if cached.  

For example, roxie can have multiple services defined and theoretically could have:

roxie:
  service1 - issuer=public, trusted_peers=client1,client2
  service1 - issuer=public, trusted_peers=client2,client3
  service1 - issuer=local, trusted_peers=client4

The trusted config with trusted peers probably doesn't need caching at all, it can just be an object held by the secure context and one level away from the secret that is checked for being stale.  But then refreshing needs to preserve the trusted peers.  Not sure if that explanation made sense, but it will come out in the wash once you dive in.

ghalliday commented:
Almost complete, but one remaining issue in dafsserver.cpp(~150) createSecureSocket overriding an option needs some thought

ghalliday commented:
@afishbeck @mckellyln @jakesmith I think this is now ready for review for logical flaws etc.  (Jake for the dafilesrv change.)
It has not been tested at all, and I plan to review it myself tomorrow before starting testing.

jakesmith reviewed (approved):
> @afishbeck @mckellyln @jakesmith I think this is now ready for review for logical flaws etc. (Jake for the dafilesrv change.) It has not been tested at all, and I plan to review it myself tomorrow before starting testing.

@ghalliday - I have only looked at the dafilesrv changes, but they look fine.

jakesmith commented:
@ghalliday - however, there's a build failure relating to an unreferenced symbol 'createSecureSocketConfig' building dafsclint
```
usr/bin/ld: CMakeFiles/dafsclient.dir/rmtclient.cpp.o: in function `_securitySettingsClient::getSecureConfig()':
/home/runner/work/HPCC-Platform/HPCC-Platform/src/fs/dafsclient/rmtclient.cpp:118: undefined reference to `createSecureSocketConfig'
collect2: error: ld returned 1 exit status
```

ghalliday commented on system/security/securesocket/securesocket.cpp:
```diff
@@ -1961,51 +1992,61 @@ extern "C" {
+        throw makeStringException(-100, "TLS secure communication requested but not configured");
+
+    Owned<const ISyncedPropertyTree> info = getIssuerTlsSyncedConfig(issuer);
+    if (!!info->isValid())
```
Oops double negative!!

ghalliday commented on system/security/securesocket/securesocket.cpp:
```diff
@@ -1961,51 +1992,61 @@ extern "C" {
 {
-    if (requireMtlsFlag && !queryMtls())
-        throw makeStringException(-100, "TLS secure communication requested but not configured");
+    if (ssf == nullptr || !ssf->queryTlsConfig())
```
could remove the second condition, since it will fall through the the same code below.

ghalliday reviewed (commented):
One bug a few typos, and some code that needs to used synced trees.

richardkchapman commented on fs/dafsclient/rmtclient.cpp:
```diff
@@ -112,6 +112,12 @@ static class _securitySettingsClient
 
+    const IPropertyTree * getSecureConfig()
+    {
+        //Later: return a synced tree...
```
When? Is that for a different PR? Is there a Jira open to revisit if so?

ghalliday replied:
HPCC-30753

richardkchapman commented on system/jlib/jptree.cpp:
```diff
@@ -10130,3 +10130,88 @@ void setExpertOpt(const char *opt, const char *value)
+//---------------------------------------------------------------------------------------------------------------------
+
+//More: Should this move inside PTree?  It may allow a more efficient hash caclulaton.
+unsigned getPropertyTreeHash(const IPropertyTree & source, unsigned hashcode)
```
Feels like this will be an expensive calculation on a large tree, that would be avoided if you could store hash of each component when it was created. But maybe it's not going to be used on large trees?

ghalliday replied:
Created HPCC-30752 to address optimizing it.  Currently it will be used on small trees.

richardkchapman commented on system/jlib/jsecrets.cpp:
```diff
@@ -1195,28 +1170,33 @@ class CSecret final : public CInterfaceOf<ISecret>
     }
+    virtual bool isValid() const override
+    {
+        CriticalBlock block(secretCs);
```
What is the point of an isValid() function like this? As soon as you release the critsec, the value you are returning may no longer be correct...

ghalliday replied:
This is typically called at startup to check that the configuration has been set up.  The calling code is cleaner since it avoids the problem with freeing the returning the linked object.

richardkchapman reviewed (commented):
Some minor comments.  A bit too large to review properly

jakesmith commented on system/jlib/jsecrets.cpp:
```diff
@@ -1364,106 +1311,233 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+
+    IPropertyTree *verify = config->queryPropTree("verify");
+    //For now only the "public" issuer implies client certificates are not required
+    verify->setPropBool("@enable", !disableMTLS && (isClientConnection || !strieq(issuer, "public")));
```
is 'verify' guaranteed to exist by this state? does it need to be null checked.

ghalliday replied:
yes - I will add an assert for safety

ghalliday replied:
yes meaning it is guaranteed to exist

jakesmith commented on system/jlib/jsecrets.cpp:
```diff
@@ -1364,106 +1311,233 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+    : CSyncedCertificateBase(nullptr), addCACert(_addCACert)
+    {
+        secret.setown(resolveSecret(_category, _secretName, nullptr, nullptr));
+        if (!secret->isValid())
```
it looks like 'secret' could be null?

ghalliday replied:
no it can't - see above.

jakesmith commented on system/security/securesocket/securesocket.cpp:
```diff
@@ -1268,9 +1268,10 @@ static bool setVerifyCertsPEMBuffer(SSL_CTX *ctx, const char *caCertBuf, int caC
 }
 
-class CSecureSocketContext : public CInterfaceOf<ISecureSocketContext>
+class CSecureSocketContext : implements ISecureSocketContext, implements ISecureSocketContextCallback, public CInterface
```
use a CInterfaceOf< to remove 1 vmt?

ghalliday replied:
I could but the IMPLEMENT_USING gets a bit ugly - not enough instances to worry either way.

jakesmith reviewed (commented):
@ghalliday - some minor comments/questions.

mckellyln commented on esp/clients/wsdfuaccess/wsdfuaccess.cpp:
```diff
@@ -543,24 +542,24 @@ StringBuffer &encodeDFUFileMeta(StringBuffer &metaInfoBlob, IPropertyTree *metaI
+        Owned<CLoadedKey> privateKey = loadPrivateKeyFromMemory(privateKeyText, nullptr);
 #else
-        privateKeyFName = environment->getPrivateKeyPath(keyPairName);
+        const char *privateKeyFName = privateKeyFName = environment->getPrivateKeyPath(keyPairName);
```
const char *privateKeyFName = privateKeyFName ?

ghalliday replied:
Interesting.  I wonder how that compiles - because the declaration comes into effect before the value is evaluated.

mckellyln commented on fs/dafsclient/rmtclient.cpp:
```diff
@@ -156,10 +162,10 @@ static ISecureSocket *createSecureSocket(ISocket *sock, const char *issuer)
-                Owned<IPropertyTree> info = getIssuerTlsServerConfig(issuer);
-                if (!info)
+                Owned<const ISyncedPropertyTree> info = getIssuerTlsSyncedConfig(issuer);
+                if (!info || !info->isValid())
```
Is it possible to check this when ptree is being built ?  And thus isValid() is true but ptree is not finished ?

mckellyln reviewed (approved):
Looks good, one question posted.
Not 100% clear to me about isValid() meaning a prop tree exists and/or a secret (if not nullptr) is present.

afishbeck commented on system/jlib/jsecrets.cpp:
```diff
@@ -1364,106 +1311,233 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+    if (!isClientConnection || !strieq(issuer, "public"))
+        updateCertificateFromSecret(secretInfo);
+
+    if (!isClientConnection || addCACert) // same condition as before, but is it correct?
```
Assuming addCACert gets set the same as before, the reason this should be correct is that addCACert is usually true.  A client hitting a public issuer is the case where we don't want the cacert defined.  Otherwise, for MTLS we want control over our CACert using addCACert.  When hitting public services using public certificate authorities we want the well known (browser compatible) CA list installed on the system instead.

ghalliday replied:
Moved your comment above into the source.

afishbeck commented on roxie/ccd/ccdmain.cpp:
```diff
@@ -1513,18 +1517,20 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
                             if (isContainerized())
                             {
+                                //MORE: Why is this containerized only?  Should it check for this in both, and fall back to the
+                                //old config if none found and non-containerized?
```
Yes, absolutely agreed, that was the ultimate goal.  Bare-metal and fusion should be able to treat certificate config as issuer secrets if they want, and if that isn't present fall back to the old style config.

ghalliday replied:
created hpcc-30754

ghalliday commented on system/jlib/jsecrets.cpp:
```diff
@@ -1364,106 +1311,233 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+    : CSyncedCertificateBase(_issuer), trustedPeers(_trustedPeers), isClientConnection(_isClientConnection), acceptSelfSigned(_acceptSelfSigned), addCACert(_addCACert), disableMTLS(_disableMTLS)
+    {
+        secret.setown(resolveSecret("certificates", issuer, nullptr, nullptr));
+        createConfig();
```
resolve secret can never return null - it always returns an object which can be used to find out if the value is defined, and may be updated behind the scenes.

ghalliday commented on system/jlib/jptree.cpp:
```diff
@@ -10130,3 +10130,88 @@ void setExpertOpt(const char *opt, const char *value)
+    return hashcode;
+}
+
+class UnchangingPropertyTree : extends CInterfaceOf<ISyncedPropertyTree>
```
I agree, changed

ghalliday commented on system/jlib/jptree.hpp:
```diff
@@ -429,4 +429,26 @@ extern jlib_decl StringBuffer &getExpertOptString(const char *opt, StringBuffer
+    virtual bool getProp(MemoryBuffer & result, const char * xpath) const = 0;
+    virtual bool getProp(StringBuffer & result, const char * xpath) const = 0;
+    //Return a sequence number which changes whenever the secret actually changes - so that a caller can determine
+    //whether it needs to reload the certificates.
```
cleaned up the comments to reflect the more general applicability

ghalliday commented on system/jlib/jptree.hpp:
```diff
@@ -429,4 +429,26 @@ extern jlib_decl StringBuffer &getExpertOptString(const char *opt, StringBuffer
+    virtual const IPropertyTree * getTree() const = 0;
+    virtual bool getProp(MemoryBuffer & result, const char * xpath) const = 0;
+    virtual bool getProp(StringBuffer & result, const char * xpath) const = 0;
+    //Return a sequence number which changes whenever the secret actually changes - so that a caller can determine
```
Clarified.

ghalliday commented on common/thorhelper/thorsoapcall.cpp:
```diff
@@ -1242,7 +1242,10 @@ class CWSCHelper : implements IWSCHelper, public CInterface
+                ownedSC.setown(createSecureSocketContextEx2(config, ClientSocket));
+            }
             else if (clientCertIssuer.length())
                 ownedSC.setown(createSecureSocketContextSecret(clientCertIssuer.str(), ClientSocket));
```
@afishbeck can you clarify this?  I think there is a problem when (certificate, key, passphrase are provided) - and I have created a jira for that.
Saving a context created from a secret globally should not matter though - because it will get updated.

afishbeck replied:
@ghalliday Yes, sorry I looked at the wrong thing.  This should be fine.

ghalliday commented:
rebased and another commit added to respond to review comments.

afishbeck commented on roxie/ccd/ccdmain.cpp:
```diff
@@ -1520,22 +1520,17 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
-                            protocol = "ssl";
 #ifdef _USE_OPENSSL
-                            if (isContainerized())
+                            protocol = "ssl";
```
This roxie config change looks good... but may want to fix line wrapping in commit message?

afishbeck commented on system/jlib/jsecrets.cpp:
```diff
@@ -1385,106 +1332,240 @@ jlib_decl bool containsEmbeddedKey(const char *certificate)
+    // addCACert is usually true. A client hitting a public issuer is the case where we don't want the ca cert
+    // defined. Otherwise, for MTLS we want control over our CACert using addCACert. When hitting public services
+    // using public certificate authorities we want the well known (browser compatible) CA list installed on the
+    // system instead.
```
Just a note, to clarify a little further:  caCert prevents a secure client from trying to use the issuer secret at all.  This is necessary because internal components don't get a public issuer secret, they don't need one.  It's ok because the OS level installed CA certs should work for public services.  

For services in this environment/namespace, config options like issuer, caCert, and selfSigned come from the list of services in our global config, for example:

```
    global:
      services:
      - caCert: true
        class: roxie
        issuer: remote
        name: roxie
        port: 29876
        public: true
        selfSigned: false
        target: roxie
        tls: true
        type: roxie 
```

ghalliday commented:
@afishbeck I have traced through the execution and fixed a few problems.  Also added some options to roxie to allow issuer in bare-metal and disable mtls.  Please take a look at the follow on commits (I think you have reviewed one of them)

ghalliday commented:
NOTE: This code has a problem- that the expiry date is stored as a property in the secret tree - so that even if the contents of the secret is identical the hash will change and the config will be updated.  Since this is once every 10 minute I think that is acceptable as a short term solution, but @afishbeck please let me know if you think otherwise.

afishbeck commented on roxie/ccd/ccdmain.cpp:
```diff
@@ -1525,7 +1525,8 @@ int CCD_API roxie_main(int argc, const char *argv[], const char * defaultYaml)
                                 certIssuer = roxieFarm.getPropBool("@public", true) ? "public" : "local";
-                            tlsConfig.setown(getIssuerTlsSyncedConfig(certIssuer, roxieFarm.queryProp("trusted_peers"), false));
+                            bool disableMtls = roxieFarm.getPropBool("@disableMtls", false);
+                            tlsConfig.setown(getIssuerTlsSyncedConfig(certIssuer, roxieFarm.queryProp("trusted_peers"), disableMtls));
```
@ghalliday Considering this is for issuer based config, I'm curious what use case / issuer scenario this is needed for?

As is, without this, public should have mtls effectively turned off, remote and local should have it turned on.
The only scenario that comes to mind is local to components in this environment, but without mtls?

Perhaps this is for a bare-metal scenario, but I'm still not sure what the details would be.

afishbeck replied:
We discussed this offline.  In bare-metal multiple components can be configured to run on the same hardware.  Different issuer names may be used to distinguish the issuer secret locations.  Therefore we may not be able to rely on the issuer name of "public".  disableMtls would be used to turn off client certificates for a public issuer that had been given a "mangled" name.

afishbeck commented:
> NOTE: This code has a problem- that the expiry date is stored as a property in the secret tree - so that even if the contents of the secret is identical the hash will change and the config will be updated. Since this is once every 10 minute I think that is acceptable as a short term solution, but @afishbeck please let me know if you think otherwise.

As we discussed offline, I don't think its a problem.  It will just be a thorough test of your secret renewal code.  Later I would optimize this away though so that the renewal only happens when something really has changed.
